app/assets/javascripts/jquery.multi-select.js

```diff
@@ -72,9 +72,41 @@ function sanitize(value){
 }
 
 $(document).on('click', '.ms-select-all', function () {
-    $(this).closest('.form-group').find('select[multiple]').multiSelect('select_all');
+  // can't use multiSelect('select_all') because it adds filtered out items too.
+    var formGroup = $(this).closest('.form-group'),
+        selected = [];
+
+    selected = itemsToSelect(formGroup);
+
+    $(formGroup).find('select[multiple]').multiSelect('select', selected);
     return false;
 });
+
+function findVisibleItems(formGroup) {
+  var visibles = [];
+  $(formGroup).find('.ms-selectable').find('ul.ms-list').children().each( function () {
```

Comment:
Why not something like `$(formGroup).find('.ms-selectable .ms-list :not(.hide)')` to find the visible items instead of this function and looping over every child? Let jQuery do what it does best - selecting stuff from the DOM.